File: src/core/sentry.py
"""
sentry.py — Sentry SDK initialization (pure function, no-op nếu DSN rỗng).

Single Responsibility: khởi tạo Sentry với đầy đủ integrations.
Được gọi 1 lần duy nhất tại server lifespan startup.
"""
from __future__ import annotations

import logging

from ..config.settings import SentryConfig

_LOG = logging.getLogger("app.sentry")


def init_sentry(cfg: SentryConfig) -> None:
    """Khởi tạo Sentry SDK. Tự động no-op nếu DSN rỗng."""
    if not cfg.dsn:
        _LOG.debug("Sentry DSN not configured — skipping initialization")
        return

    import sentry_sdk
    from sentry_sdk.integrations.asyncio import AsyncioIntegration
    from sentry_sdk.integrations.fastapi import FastApiIntegration
    from sentry_sdk.integrations.logging import LoggingIntegration
    from sentry_sdk.integrations.starlette import StarletteIntegration

    sentry_sdk.init(
        dsn=cfg.dsn,
        environment=cfg.environment,
        integrations=[
            StarletteIntegration(transaction_style="endpoint"),
            FastApiIntegration(transaction_style="endpoint"),
            AsyncioIntegration(),
            LoggingIntegration(
                level=logging.WARNING,   # breadcrumb từ WARNING trở lên
                event_level=logging.ERROR,  # tạo Sentry event từ ERROR trở lên
            ),
        ],
        traces_sample_rate=cfg.traces_sample_rate,
        profiles_sample_rate=cfg.profiles_sample_rate,
        send_default_pii=cfg.send_default_pii,
    )
    _LOG.info("Sentry initialized (env=%s, traces=%.2f)", cfg.environment, cfg.traces_sample_rate)

File: ui/src/lib/sentry.ts
/**
 * sentry.ts — Sentry SDK initialization for React UI.
 * No-op nếu DSN không được cấu hình (VITE_SENTRY_DSN rỗng).
 */
import * as Sentry from "@sentry/react";

const DSN = "https://02e1b958d2e9226ac579eb96a3be5b1a@o4511148563890176.ingest.us.sentry.io/4511148594954240";
const ENV = import.meta.env.VITE_SENTRY_ENV ?? "production";

export const initSentry = (): void => {
  if (!DSN) return;

  Sentry.init({
    dsn: DSN,
    environment: ENV,
    integrations: [
      Sentry.browserTracingIntegration(),
      Sentry.replayIntegration({
        maskAllText: true,
        blockAllMedia: true,
      }),
    ],
    tracesSampleRate: 0.1,
    replaysSessionSampleRate: 0.0,
    replaysOnErrorSampleRate: 1.0,
  });
};

export { Sentry };
